onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -0,0 +1,276 @@
+
+// Example 2: current_dims = [0, 0, 0] tensor_dims = [1, 2, 2], then current_dims = [0, 1, 0]
+//          current_dims = [0, 1, 0] tensor_dims = [1, 2, 2], then current_dims = [0, 0, 0]
+void increment_over_inner_dim(std::vector<int64_t>& current_dims, const std::vector<int64_t>& tensor_dims) {
+  // in this context, rank can never be < 1, so saving checking overhead
+  int64_t rank = static_cast<int64_t>(current_dims.size());
+
+  // 'reset' innermost dimension value
+  current_dims[0] = 0;
+
+  // nothing to increment over
+  if (rank == 1) {
+    return;
+  }
+
+  int64_t current_axis = rank - 2;
+
+  while (current_axis >= 0) {
+    if (++current_dims[current_axis] != tensor_dims[current_axis])
+      return;
+
+    current_dims[current_axis] = 0;
+    --current_axis;
+  }
+}
+
+// parse indices_tensor and along the way validate its shape and contents
+std::vector<int64_t> parse_and_validate_indices_tensor(const Tensor* indices_tensor,
+                                                             int64_t axis, const std::vector<int64_t>& input_shape) {
+  const auto& indices_shape = indices_tensor->Shape().GetDims();
+  int64_t indices_rank = static_cast<int64_t>(indices_shape.size());
+  for (int64_t i = 0; i < indices_rank; ++i) {
+    // for all axes except the axis of interest,
+    // make sure that the corresponding 'indices' shape
+    // value if within bounds of the corresponding 'data' shape
+    if (i != axis) {
+      if (indices_shape[i] < 0 || indices_shape[i] > indices_shape[i])
+        ORT_THROW("GatherElements op: 'indices' shape should have values within bounds of 'data' shape");
+    }
+  }
+
+  // first parse 'indices' data
+  auto num_elements = indices_tensor->Shape().Size();
+  std::vector<int64_t> indices_data;
+  // reserving memory ahead as we know the size of the container
+  indices_data.reserve(num_elements);
+  if (indices_tensor->DataType() == DataTypeImpl::GetType<int32_t>()) {
+    const auto* data = indices_tensor->Data<int32_t>();
+    for (int64_t i = 0; i < num_elements; ++i)
+      indices_data.push_back(data[i]);
+  } else if (indices_tensor->DataType() == DataTypeImpl::GetType<int64_t>()) {
+    const auto* data = indices_tensor->Data<int64_t>();
+    for (int64_t i = 0; i < num_elements; ++i)
+      indices_data.push_back(data[i]);
+  } else {
+    ORT_THROW("GatherElements op: Data type for 'indices' tensor must be 'int32_t' and 'int64_t'");
+  }
+
+  // validate 'indices' data
+  // along the way 'fix' negative index values if within bounds
+  int64_t lower_index_limit = -input_shape[axis];
+  int64_t upper_index_limit = input_shape[axis] - 1;
+
+  for (int64_t i = 0; i < num_elements; ++i) {
+    if (indices_data[i] < lower_index_limit || indices_data[i] > upper_index_limit)
+      ORT_THROW("GatherElements op: Value in indices must be within bounds [",
+                lower_index_limit, " , ", upper_index_limit, "]");
```

Comment:
Nicer to include the invalid value as well.

Reply:
Sure.

Reply:
Fixed.